### DISCUSSION TITLE
Dialect not properly initialized

### DISCUSSION THREAD

**Discussion by johnraz** [sqlalchemy/sqlalchemy#5992] [Usage Questions]
**Describe the bug**
For a reason I couldn't define so far, the dialect is sometimes wrongly initialized leading to eg: `engine.dialect._has_native_json` being set to `False` instead of `True` using Postgresql and the latest version of pyscopg2.

More than often and without any identifiable pattern the deserialization error shown below happens.

which I tracked to be exactly the same root cause as described in https://github.com/sqlalchemy/sqlalchemy/issues/4807

--> the Dialect is not initialized properly or more likely not initialized at all which leads to JSON fields being deserialized twice --> the second attempt raises the error.

**Expected behavior**

The dialect should be properly initialized and the json field only deserialized once.

**To Reproduce**
I wasn't able to reproduce this with a test case, sadly :(

**Error**

```
TypeError: the JSON object must be str, bytes or bytearray, not 'dict'

  ... I removed some proprietary code traceback here but it's basically accessing some related field on a model...

  File "sqlalchemy/orm/attributes.py", line 294, in __get__
    return self.impl.get(instance_state(instance), dict_)
  File "sqlalchemy/orm/attributes.py", line 730, in get
    value = self.callable_(state, passive)
  File "sqlalchemy/orm/strategies.py", line 760, in _load_for_state
    session, state, primary_key_identity, passive
  File "<string>", line 1, in <lambda>
  File "sqlalchemy/orm/strategies.py", line 902, in _emit_lazyload
    .with_post_criteria(set_default_params)
  File "sqlalchemy/ext/baked.py", line 544, in all
    return list(self)
  File "sqlalchemy/orm/loading.py", line 100, in instances
    cursor.close()
  File "sqlalchemy/util/langhelpers.py", line 70, in __exit__
    with_traceback=exc_tb,
  File "sqlalchemy/util/compat.py", line 182, in raise_
    raise exception
  File "sqlalchemy/orm/loading.py", line 80, in instances
    rows = [proc(row) for row in fetch]
  File "sqlalchemy/orm/loading.py", line 80, in <listcomp>
    rows = [proc(row) for row in fetch]
  File "sqlalchemy/orm/loading.py", line 588, in _instance
    populators,
  File "sqlalchemy/orm/loading.py", line 725, in _populate_full
    dict_[key] = getter(row)
  File "sqlalchemy/sql/type_api.py", line 1275, in process
    return process_value(impl_processor(value), dialect)
  File "sqlalchemy/sql/sqltypes.py", line 2454, in process
    return json_deserializer(value)
  File "json/__init__.py", line 348, in loads
    'not {!r}'.format(s.__class__.__name__))
```

**Versions.**
 - OS: linux
 - Python: 3.6.12
 - SQLAlchemy: 1.3.23
 - Database: Posgresql 11.7
 - DBAPI: 2.8.6

**Additional context**

In our case the bug triggers in an [Dramatiq](https://dramatiq.io/) setup with a lot of tasks running.

We use [flask-sqlalchemy](https://flask-sqlalchemy.palletsprojects.com/) in combinaison with [flask-melodramatiq](https://flask-melodramatiq.readthedocs.io/).

The two *should*  play well together and the lifespan of a session should be the lifespan of the app context which is created when a task run and disposed of when the tasks is done executing.

Tracing the error in the logs show really no pattern of any kind, it happens on different tasks / different models / different ids and different machines on our cluster.

I also cannot reproduce locally which is making the all thing even harder to diagnose ^^

I have spotted in our codebase an abusive use of `session.bind.dispose()`  in an attempt to limit the number of connections to the DB while a simple `session.commit()` / `session.rollback()` should be enough - I removed that and I'm waiting for a release in the coming days to see if it solved anything - but I figured I would mention it anyway

What I'm looking for now is suggestions or clues of where this initialization might fail in order to narrow an pinpoint the source of the issue.

Thanks for your help 👍

**Comment by johnraz** [comment]
@zzzeek I'm taking the liberty to ping you here as you were handling the related issue 🙏

**Comment by zzzeek** [comment]
has_native_json is determined only when the engine first connects.  it is based on the psycopg2 version, however that's when it gets checked.    you can set it manually to True.   Also, _has_native_json is removed in 1.4 as it is always true for the current supported versions of psycopg2.

**Reply by johnraz** [reply]
Thanks for the feedback.

I have figured all that and checked thoroughly and I can't seem to find an explanation as to why `_has_native_json` would be badly initialized (pyscopg version etc are fine) or left to the default value of `False`.

I could go ahead and  manually set it to True but my concern is more that the engine seems to end up in the middle of a query improperly initialized  already and that allowing it to move forward with the query would cause even more troubles?

**Reply by zzzeek** [reply]
without a reproducer it's hard to know where to go here.  check that you don't have an old version of psycopg2 being installed, and that there is no manipulation of dialects outside of the usual create_engine() routine.

**Reply by johnraz** [reply]
Yup I'm 100% positive pyscopg2 is the last version and dialects are not overridden.

I just added some more logs to try to find out what's happening but I have the feeling that it's some kind of concurrency issue maybe some kind of thread unsafety at some point... but as you said very hard to move forward without a reproducer.

I'll keep digging and will keep you posted if I find anything worth sharing.

Thanks again for your time.

**Reply by zzzeek** [reply]
OK I see you've mentioned #4807 but that adds a mutex that is trying to guard against this condition.  if the "exec once" state, which is where this initialize happens, hasnt completed, everyone should be blocking on that mutex.   so if that's not working then there's a bug.

**Reply by zzzeek** [reply]
here's a POC for #4807

```
from sqlalchemy import create_engine
import threading


e = create_engine("postgresql://scott:tiger@localhost/test", echo='debug')


def run():
    print("worker starting....")
    with e.connect() as conn:
        print("worker connected!!")
        print(conn.execute("select 1").scalar())
    print("worker done")


threads = [threading.Thread(target=run) for i in range(10)]
for t in threads:
    t.start()
for t in threads:
    t.join()


```

I run this while applying this patch:

```diff
diff --git a/lib/sqlalchemy/dialects/postgresql/psycopg2.py b/lib/sqlalchemy/dialects/postgresql/psycopg2.py
index 2c5a934db..f81d7c37c 100644
--- a/lib/sqlalchemy/dialects/postgresql/psycopg2.py
+++ b/lib/sqlalchemy/dialects/postgresql/psycopg2.py
@@ -754,6 +754,8 @@ class PGDialect_psycopg2(PGDialect):
                 )
 
     def initialize(self, connection):
+        import time
+        time.sleep(5)
         super(PGDialect_psycopg2, self).initialize(connection)
         self._has_native_hstore = (
             self.use_native_hstore

```

when I run it, all ten connections do not complete for five seconds:

```
$ python test3.py 
worker starting....
worker starting....
worker starting....
worker starting....
worker starting....
worker starting....
worker starting....
worker starting....
worker starting....
worker starting....
<waits 5 seconds here>

```
then they all get through at once:
```

2021-03-01 13:14:30,403 INFO sqlalchemy.engine.base.Engine select version()
2021-03-01 13:14:30,404 INFO sqlalchemy.engine.base.Engine {}
2021-03-01 13:14:30,406 DEBUG sqlalchemy.engine.base.Engine Col ('version',)
2021-03-01 13:14:30,407 DEBUG sqlalchemy.engine.base.Engine Row ('PostgreSQL 12.4 on x86_64-redhat-linux-gnu, compiled by gcc (GCC) 10.2.1 20200826 (Red Hat 10.2.1-3), 64-bit',)
2021-03-01 13:14:30,409 INFO sqlalchemy.engine.base.Engine select current_schema()
2021-03-01 13:14:30,409 INFO sqlalchemy.engine.base.Engine {}
2021-03-01 13:14:30,410 DEBUG sqlalchemy.engine.base.Engine Col ('current_schema',)
2021-03-01 13:14:30,411 DEBUG sqlalchemy.engine.base.Engine Row ('public',)
2021-03-01 13:14:30,413 INFO sqlalchemy.engine.base.Engine SELECT CAST('test plain returns' AS VARCHAR(60)) AS anon_1
2021-03-01 13:14:30,413 INFO sqlalchemy.engine.base.Engine {}
2021-03-01 13:14:30,415 INFO sqlalchemy.engine.base.Engine SELECT CAST('test unicode returns' AS VARCHAR(60)) AS anon_1
2021-03-01 13:14:30,416 INFO sqlalchemy.engine.base.Engine {}
2021-03-01 13:14:30,417 INFO sqlalchemy.engine.base.Engine show standard_conforming_strings
2021-03-01 13:14:30,417 INFO sqlalchemy.engine.base.Engine {}
2021-03-01 13:14:30,418 DEBUG sqlalchemy.engine.base.Engine Col ('standard_conforming_strings',)
2021-03-01 13:14:30,418 DEBUG sqlalchemy.engine.base.Engine Row ('on',)
worker connected!!
worker connected!!
worker connected!!
2021-03-01 13:14:30,421 INFO sqlalchemy.engine.base.Engine select 1
worker connected!!
worker connected!!
worker connected!!
worker connected!!
... etc.
```

.  so the init of he first connect is pausing all the other connections successfully.

**Reply by johnraz** [reply]
My gut feeling is that something is really off, like a bug indeed.
I went down most of the path I could think off and this is the only theory I'm left with.

That's why I was trying to narrow other areas where could overridden or worked around or something.

I probably need to dig more into the SQLA source myself to get a better understanding of that mutex.

**Reply by johnraz** [reply]
Thanks for the POC - will try to learn from that tomorrow 👍

**Reply by johnraz** [reply]
Could the `initialize` method call be skipped by any chance? Would some code path allow that?

**Reply by zzzeek** [reply]
not if you aren't doing anything unusual like sharing the pool among engines or something.

**Reply by zzzeek** [reply]
I would advise patching some logging or print statements into the dialect's initialize routine, combine that with [engine logging](https://docs.sqlalchemy.org/en/13/core/engines.html#configuring-logging) using the Python logging interface.  then use a [formatter that includes the thread id](https://javawithravi.com/how-to-display-thread-id-in-python-logs/).  that way you can see exactly if any threads are running when they shouldnt.

**Reply by johnraz** [reply]
Cool that's a very neat idea - will add some more logging 👍

**Reply by johnraz** [reply]
So I've been running your POC:

- It runs as expected in my docker image so same python, SQLA and psycopg then in our prod.
- I added the flask app in the POC and removed the sleep --> queried the ORM, same result everything works as expected.

I had some hope that I could reproduce this way but nope :(

I will move on with your logging suggestion.

Although can you elaborate on what you mean by:

> that way you can see exactly if any threads are running when they shouldnt

I'm not 100% sure what I should be looking at specifically?

**Reply by zzzeek** [reply]
here's what good log output would look like:

```
Thread-1 - Starting initialize!
Thread-1 - Doing initialize things!
Thread-1 - Done initializing
Thread-1 - Connected!
Thread-1 - running a query!
Thread-2 - Connected!
Thread-3 - Connected!
Thread-3 - Running a query!
Thread-2 - Running a query!

```

Here's bad output:


```
Thread-1 - Starting initialize!
Thread-1 - Doing initialize things!
Thread-2 - Connected!
Thread-2 - Running a query!
Thread-3 - Connected!
Thread-1 - Done initializing
Thread-1 - Connected!
Thread-1 - running a query!
Thread-3 - Running a query!

```


the idea is that when logging indicates the "init" is occurring, from start to end, you should see no other threads happening.   

i think @caselit 's notions of forking / subprocesses might be a factor in what you are seeing.    If anything is running os.fork() /subprocess() / etc. while the lead process is doing this initialize, then that's something that might lead to this issue.  so you want to have logging that also includes process ids, not just thread ids, if your architecture is running multiple python processes spawned off from the lead process.

**Comment by CaselIT** [comment]
Not sure if maybe trying to add a dramatiq middleware that implements `after_process_boot` may be useful to initialize (or dispose) the engine before any usage? You could also force initialization issuing a `select 1;` like query to the db

**Reply by johnraz** [reply]
@CaselIT Thanks for the suggestion; this is what I did today; adding a `select true;` query  in the `before_process_message` and I should probably a commit / rollback at `after_process_message`.

I'm not sure to follow how it would be needed at the `process` level?
Considering that the Flask context / SQLAlchemy session is instantiated / disposed with the message cycle, the outer scope shouldn't matter right?

Or Am I missing a use case where that should matter?

Thanks again for the invaluable help 👍

**Reply by CaselIT** [reply]
I suggested after process boot (but I think also before/after worker boot may be the same) to avoid the chance of possibly using an engine initialized in another process. Usually the errors are different but it may be worth a try
